crypto: generate_config_checks.py: use current data

gilles-peskine-arm opened the pull request:
Use current data rather than data from 1.0 to generate config checks. Fix https://github.com/Mbed-TLS/mbedtls-framework/issues/249

Needs preceding PR: submodule in https://github.com/Mbed-TLS/mbedtls-framework/pull/260

## PR checklist

- [x] **changelog** not required because: not user-visible
- [x] **framework PR** provided https://github.com/Mbed-TLS/mbedtls-framework/pull/260
- [x] **crypto PR** provided https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/619
- [x] **mbedtls development PR** provided https://github.com/Mbed-TLS/mbedtls/pull/10548
- [x] **mbedtls 3.6 PR** not required because: feature added in 1.0/4.0
- **tests**  provided

gilles-peskine-arm committed:
Update framework with config_macros.py

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Switch from config_history to config_macros

Switch from the `config_history` module to the new module `config_macros`.
No behavior change.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Read current data rather than data about 1.0

It doesn't matter how a macro was used in a previous minor version of the
library. What matters is current information about options and internal
symbols, and information about past versions from which a macro may have
been removed.

The output is mostly the same, but:

* Macros that were options in 3.6, became internal in 1.0 and have now
  been completely removed are now shown as removed, not internal.
* Macros that were options in 3.6, were completely removed in 1.0, and are
  now back but internal, are now shown as internal, not removed.
* Macros that were options in 3.6, were removed in 1.0 and are back to
  being options are no longer rejected.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
If an option was removed, always reject it as removed

If an option was removed, reject it as removed, whether or not it is also
internal. For long-time users, this is more informative. For new users who
were not aware of the macro's previous use as a public option, the extra
information is not confusing.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Also reject internal macros that weren't options in 3.6

But allow `MBEDTLS_PSA_ACCEL_` in the config file.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
New script to check or update config-options-current.txt

This script may be generalized to check other files that need lists of
current options, such as files that declare or check `MBEDTLS_PSA_ACCEL_xxx`
or `MBEDTLS_PSA_BUILTIN_xxx` for (almost) all available mechanisms. But for
now, the script just checks `scripts/data_files/config-options-current.txt`.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Create the config option list shadow file

```
tests/scripts/check_option_lists.py -u
```

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm committed:
Check that the config option list shadow file is up to date

Test that `scripts/data_files/config-options-current.txt` is up-to-date.
This file needs to change every time we add or remove a config option.

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

gilles-peskine-arm commented on scripts/data_files/config-options-current.txt:
```diff
@@ -0,0 +1,219 @@
+MBEDTLS_AESCE_C
```
To regenerate this file, run
```
tests/scripts/check_option_lists.py -u
```

This command appears in the commit message that adds the file. It doesn't appear in the file itself because I wanted to keep it simple, so there's no room for a comment.

gilles-peskine-arm commented on scripts/data_files/config-options-current.txt:
```diff
@@ -0,0 +1,219 @@
+MBEDTLS_THREADING_ALT
+MBEDTLS_THREADING_C
+MBEDTLS_THREADING_PTHREAD
+PSA_CRYPTO_CONFIG_H
```
The presence of the header guard is preexisting. I didn't consider it a problem in https://github.com/Mbed-TLS/mbedtls-framework/pull/196 and I still don't. It has very little practical consequence: users will be allowed to define `PSA_CRYPTO_CONFIG_H` in their crypto config file, which is exactly what would happen if the macro was omitted when constructing the list of options. In Mbed TLS, users will not be allowed to define `PSA_CRYPTO_CONFIG_H` if the crypto config file doesn't define it, which is actually kind of desirable even though it doesn't really matter.

mpg reviewed (dismissed):
LGTM, thanks!

gilles-peskine-arm committed:
Print a useful message when the file is out of date

Don't print the differences: interested users can just run `git diff` (or
save the old file and run `comm`).

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

mpg commented on tests/scripts/check_option_lists.py:
```diff
@@ -35,7 +35,11 @@ def main():
+        if not data.is_shadow_file_up_to_date():
+            print(f'{data.shadow_file_path()} is out of date')
+            print(f'After adding or removing a conifg option, you need to run')
+            print(f'{sys.argv[0]} and commit the result.')
```
Shouldn't we suggest the `-u` option? Otherwise, the script I'm running (presumably without options) is telling me to run the same script again (apparently without options)...

mpg commented on tests/scripts/check_option_lists.py:
```diff
@@ -35,7 +35,11 @@ def main():
-        up_to_date &= data.compare_shadow_file_verbosely()
+        if not data.is_shadow_file_up_to_date():
+            print(f'{data.shadow_file_path()} is out of date')
+            print(f'After adding or removing a conifg option, you need to run')
```
typo: conIFg

mpg reviewed (changes requested):
LGTM overall, but the output message can be improved.

gilles-peskine-arm committed:
Improve out-of-date message

Signed-off-by: Gilles Peskine <Gilles.Peskine@arm.com>

mpg reviewed (dismissed):
LGTM, thanks!

gilles-peskine-arm commented:
I rebased on top of the target branch to resolve the conflict with the framework submodule, and updated the submodule-updating commit accordingly. The previous history is at https://github.com/gilles-peskine-arm/TF-PSA-Crypto/tree/config-checks-generator-current-crypto-6

mpg reviewed (approved):
LGTM, thanks!
